### ISSUE TITLE
APM model can cause deadlock under TPL reentrancy optimizations

### ISSUE THREAD

**Issue by danielmarbach** [Azure/azure-storage-net#378]
@Scooletz @tmasternak just spend several hours debugging a deadlock issue we found on our CI environment. The TLDR is the following:

* The APM model the Azure Storage Queue async calls are using can cause deadlocks under resource restricted environments (in our case two cores on Amazon T2 large) when the TPL runtime tries to optimize code with reentrancy due to reusing of threads. You need to have a pure async calling site.
* There are two potential fixes:
  * Option 1: All the `TaskCompletionSources` (i.ex. [here](https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Core/Util/AsyncExtensions.cs#L153)) used in the [`AsyncExtensions`](https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Core/Util/AsyncExtensions.cs) need to specify the [`TaskCreationOptions.RunContinuationsAsynchronously`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskcreationoptions.aspx), this requires framework reference > 4.6 which means the nuspec has to be updated as well. _or_
  * Option 2: The `CreateCallback` method needs to execute the lambda which calls `TrySetResult`, `TrySetCanceled` and so on inside a `Task.Run`, doesn't require a .NET Framework bump

Happy to provide a PR for option 1 or option 2. We have successfully tested option 1 on our CI environments but not option 2. 

## Additional information

In our case [`GetMessagesAsync` ](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/develop/src/Transport/AzureMessageQueueReceiver.cs#L56)was deadlocking. Here is our trace output

```
After getmessgesasync for thread 7 for queue unsubscribingfromevent-publisher
Subscribe from UnsubscribingFromEvent.Subscriber2 on message type NServiceBus.AcceptanceTests.Routing.MessageDrivenSubscriptions.When_unsubscribing_from_event+Event, NServiceBus.Azure.Transports.WindowsAzureStorageQueues.AcceptanceTests, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
Before getmessgesasync for thread 11 for queue unsubscribingfromevent-publisher
After getmessgesasync for thread 21 for queue unsubscribingfromevent-subscriber2-retries
After getmessgesasync for thread 7 for queue unsubscribingfromevent-subscriber2
After getmessgesasync for thread 21 for queue unsubscribingfromevent-publisher-retries
After getmessgesasync for thread 20 for queue unsubscribingfromevent-publisher
P: Event published
* Context: Subscriber1Subscribed: False, Subscriber2Subscribed: True, Subscriber2Unsubscribed: False, Subscriber1ReceivedMessages: 0, Subscriber2ReceivedMessages: 0
Before getmessgesasync for thread 28 for queue unsubscribingfromevent-publisher
Before getmessgesasync for thread 28 for queue unsubscribingfromevent-publisher-retries
Before getmessgesasync for thread 28 for queue unsubscribingfromevent-subscriber2
Before getmessgesasync for thread 28 for queue unsubscribingfromevent-subscriber2-retries
```
After that our message pump was completely dead locked. An explanation might be:

Under our pure async code the thread 28 in the above example is executing scheduled tasks of multiple of our message pumps. The BeginGetMessages part is synchronously executed, when the EndCall comes back it tries to set the TaskCompletionSource which then triggers the continuation to be executed on thread 28 which can cause deadlocks when other async operations where scheduled on thread 28.

If we put a `Thread.Sleep` into the right place of our async acceptance testing framework then we can block the thread that would enter normally into GetMessagesAsync and enforcing a thread switch. Then the deadlock will not occur.

```
 After getmessgesasync for thread 23 for queue unsubscribingfromevent-subscriber2-retries
 After getmessgesasync for thread 23 for queue unsubscribingfromevent-publisher
 Before getmessgesasync for thread 11 for queue unsubscribingfromevent-publisher-retries
 After getmessgesasync for thread 23 for queue unsubscribingfromevent-publisher-retries
 Before getmessgesasync for thread 22 for queue unsubscribingfromevent-subscriber2
 P: Event published
 After getmessgesasync for thread 10 for queue unsubscribingfromevent-subscriber2
 S2: event received
 Before getmessgesasync for thread 11 for queue unsubscribingfromevent-subscriber2
 S2: Sending unsubscribe
 After getmessgesasync for thread 10 for queue unsubscribingfromevent-subscriber2
 S2: Unsubscription sent
 Before getmessgesasync for thread 22 for queue unsubscribingfromevent-publisher
 Before getmessgesasync for thread 22 for queue unsubscribingfromevent-subscriber2-retries
 After getmessgesasync for thread 16 for queue unsubscribingfromevent-subscriber2-retries
 Before getmessgesasync for thread 22 for queue unsubscribingfromevent-publisher-retries
 After getmessgesasync for thread 24 for queue unsubscribingfromevent-publisher-retries
 Before getmessgesasync for thread 11 for queue unsubscribingfromevent-subscriber2
 After getmessgesasync for thread 16 for queue unsubscribingfromevent-subscriber2
 Before getmessgesasync for thread 12 for queue unsubscribingfromevent-publisher-retries
 Before getmessgesasync for thread 22 for queue unsubscribingfromevent-subscriber2-retries
 After getmessgesasync for thread 24 for queue unsubscribingfromevent-publisher-retries
```

**Comment by asorrin-msft** [comment]
Daniel - 
We've been looking at this, and we're confused as to where the deadlock could be coming from.  Do you have a stack trace we could examine, showing where the threads were / what they were waiting on when the application deadlocked?

I know about the sorts of synchronization-context related deadlocks you can get when you try to [mix sync and async code](https://blogs.msdn.microsoft.com/pfxteam/2012/04/13/should-i-expose-synchronous-wrappers-for-asynchronous-methods/), but I don't think that could be happening here, as neither our code nor yours is synchronously waiting.  Otherwise, I don't see where the deadlock could be coming from.  
From (briefly) browsing the .NET reference source, it does look like TaskCompletionSource.TrySetResult will run the continuation synchronously, but that's synchronously on the thread on which the AsyncCallback (the callback passed into the Begin method) was called, which is not (necessarily) the thread on which Begin was called in the first place, and thus shouldn't be blocking anywhere.  (Unless I'm mistaken somewhere - if so, please let me know!)  If the AsyncCallback is getting called synchronously with the thread that called Begin, this also shouldn't cause a deadlock, because that means the thread was never doing anything else and was available to process the continuation.  It's certainly possible there's something else going on (you mentioned the TPL runtime trying to optimize thread use, not sure what that means), but if so I'd love to understand what it is.

The problem I have with both of your suggestions is that they're both removing optimizations that would otherwise be good.  If there's a correctness issue, then we should definitely fix it, but I would like to completely understand where the deadlock is coming from.

Thanks!  (And thanks for the other two fixes, we'll be accepting those shortly.)

**Comment by danielmarbach** [comment]
Thanks for coming back to us! Highly appreciated. We might need to dive our heads again into the issue and go through the source code again. My memory is garbage collected after New Year's Eve ;) Will consult with @Scooletz @tmasternak and update the issue. In the meantime here are a few interesting readings

https://github.com/dotnet/roslyn/commit/85c4c37e9d79ce04cb3b73cad31926d6c3bef1f5
http://stackoverflow.com/questions/37087103/wcf-client-hangs-on-any-operation-after-awaiting-async-operation-in-console-appl
http://stackoverflow.com/questions/12245935/is-task-factory-startnew-guaranteed-to-use-another-thread-than-the-calling-thr
https://github.com/dotnet/corefx/issues/2454

Could it be that the client has WaitHandle.Wait or a lock somewhere which could lead to that issue under restricted environment with contention?

**Comment by danielmarbach** [comment]
Hi @asorrin-msft 

@scooletz and I took another round of brainstorming and just before our head exploded we stopped, and that's why we are coming back to you. 

We followed your reasoning and went through ours and your code path again. We agree that nothing is blocking in your framework or ours that could introduce this problem. We can consistently reproduce the hang (let's not call it deadlock) on our CI environments. We would appreciate any help and brainstorming on this matter. Here are the stack traces from the hanging thread/s

[StackTraceThread26-When.txt](https://github.com/Azure/azure-storage-net/files/716627/StackTraceThread26-When.txt)

What we tried as well is to change the GetMessagesAsync method to pass in the following options.

`new QueueRequestOptions { MaximumExecutionTime = TimeSpan.FromSeconds(5), ServerTimeout = TimeSpan.FromSeconds(5)}`

not even that resolves it. Here is the full stack trace of the whole process with the above options set.

[FullStackDump.txt](https://github.com/Azure/azure-storage-net/files/716628/FullStackDump.txt)

What was a bit suspicious to us was the [`void AsyncStreamCopier<T>.StartCopyStream(Action<ExecutionState<T>> completedDelegate, long? copyLength, long? maxLength)
`](https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Core/Util/AsyncStreamCopier.cs#L85) that internally kicks of a task. This method calls into [`StartCopyStreamAsync`](https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Core/Util/AsyncStreamCopier.cs#L201) which has a [lock statement ](https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Core/Util/AsyncStreamCopier.cs#L206) that seems to be happening in the synchronous part of the async method. We are not sure if it is related to the lock, so it is just a hint. We consistently see our code hanging around this AsyncStreamCopier.

Thanks again for your involvement on this matter.

**Comment by danielmarbach** [comment]
@asorrin-msft After that issue started happening with our ASB transport as well we were able to track it down to the following code of ourshttps://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.AcceptanceTesting/Support/EndpointRunner.cs#L109. The too tight loop and the fact that the when condition checking most likely will happen on the synchronous path introduced a starvation problem. The following pull addresses this and already made our tests green. 

https://github.com/Particular/NServiceBus/pull/4437

Closing this
